File: src/modules/Manager.ts
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
import { MessageId, Messages, VoidCallback, _SystemMessages, _ID_MESSAGES } from "./modules_const";
import { ID_YANDEX_METRICA, _UserMessages } from "../main/game_config";
import { register_system } from "./System";
import { register_log } from "./Log";
import { register_event_bus } from "./EventBus";
import { register_storage } from "./Storage";
import { register_game_storage } from "./GameStorage";
import { register_sound } from "./Sound";
import { register_lang } from "./Lang";
import { register_scene } from "./Scene";
import { register_ads } from "./Ads";
import { register_rate } from "./Rate";
import { register_metrica } from "./Metrica";
import { register_camera } from "./Camera";
import { register_html_bridge } from './HtmlBridge';



/*
    Основной модуль для подгрузки остальных, доступен по объекту Manager
    также глобально доступна функция to_hash которая ограничит список доступных для отправки сообщений
    при проверке в on_message, например  if (message_id == to_hash('MANAGER_READY'))
*/

declare global {
    const Manager: ReturnType<typeof ManagerModule>;
    const to_hash: typeof _to_hash;
    type UserMessages = _UserMessages;
    type SystemMessages = _SystemMessages;
    const ID_MESSAGES: typeof _ID_MESSAGES;
}

const hashed_keys: { [key: string]: hash } = {};
function _to_hash<T extends MessageId>(key: T) {
    if (hashed_keys[key] != undefined)
        return hashed_keys[key];
    hashed_keys[key] = hash(key);
    return hashed_keys[key];
}

export function register_manager() {
    (_G as any).Manager = ManagerModule();
    (_G as any).to_hash = _to_hash;
    (_G as any).ID_MESSAGES = _ID_MESSAGES;
}


function ManagerModule() {
    let _is_ready = false;
    const MANAGER_ID = 'main:/manager';
    const UI_ID = '/ui#game';
    const LOGIC_ID = '/game_logic#game';
    const VIEW_ID = '/game_view#view';

    function init(callback_ready?: VoidCallback, use_custom_storage_key = false) {
        math.randomseed(socket.gettime());
        register_system();
        register_log();
        register_storage(use_custom_storage_key);
        register_game_storage();
        register_event_bus();
        if (System.platform == 'HTML5') {
            register_html_bridge();
            HtmlBridge.init(() => register_modules(callback_ready));
        }
        else
            register_modules(callback_ready);
    }

    function register_modules(callback_ready?: VoidCallback) {
        register_metrica();
        register_sound();
        register_lang();
        register_scene();
        register_camera();
        register_ads();
        register_rate();
        Metrica.init(ID_YANDEX_METRICA);
        check_ready(callback_ready);
    }


    function check_ready(callback_ready?: VoidCallback) {
        const id_timer = timer.delay(0.1, true, () => {
            // список модулей с ожиданием готовности
            if (Ads.is_ready()) {
                timer.cancel(id_timer);
                _is_ready = true;
                log('All Managers ready ver: ' + sys.get_config("project.version"));
                send('MANAGER_READY');
                if (callback_ready)
                    callback_ready();
            }
        });
    }

    function is_ready() {
        return _is_ready;
    }

    // отправить сообщение
    function send_raw(message_id: string | hash, message_data?: any, receiver = MANAGER_ID) {
        msg.post(receiver, message_id, message_data);
    }

    function send_raw_ui(message_id: string | hash, message_data?: any, receiver = UI_ID) {
        send_raw(message_id, message_data, receiver);
    }

    function send_raw_view(message_id: string | hash, message_data?: any, receiver = VIEW_ID) {
        send_raw(message_id, message_data, receiver);
    }

    function send_view<T extends MessageId>(message_id: T, message_data?: Messages[T], receiver = VIEW_ID) {
        send_raw(message_id, message_data, receiver);
    }

    function send_raw_game(message_id: string | hash, message_data?: any, receiver = LOGIC_ID) {
        send_raw(message_id, message_data, receiver);
    }

    function send_game<T extends MessageId>(message_id: T, message_data?: Messages[T], receiver = LOGIC_ID) {
        send_raw(message_id, message_data, receiver);
    }

    function send<T extends MessageId>(message_id: T, message_data?: Messages[T], receiver = MANAGER_ID) {
        send_raw(message_id, message_data, receiver);
    }

    function on_message(_this: any, message_id: hash, message: any, sender: hash) {
        EventBus.on_message(_this, message_id, message, sender);
    }

    function on_message_main(_this: any, message_id: hash, message: any, sender: hash) {
    	if (_is_ready){
        	Scene._on_message(_this, message_id, message, sender);
        	Sound._on_message(_this, message_id, message, sender);
        	Rate._on_message(_this, message_id, message, sender);
    	}
    }

    // можно вызывать в каждом init всех gui/go чтобы применялись языки например
    function init_script() {
        Lang.apply();
        EventBus.on('ON_APPLY_CUSTOM_LANG', () => Lang.apply());
    }

    function final_script() {
        EventBus.off_all_current_script();
    }


    return {
        init, on_message_main, on_message, is_ready, init_script, final_script, send, send_raw, send_game, send_raw_game, send_view, send_raw_view, send_raw_ui,
        MANAGER_ID, UI_ID, LOGIC_ID, VIEW_ID
    };
}

File: app/modules/Manager.lua
local ____exports = {}
local ManagerModule
local ____modules_const = require("modules.modules_const")
local _ID_MESSAGES = ____modules_const._ID_MESSAGES
local ____game_config = require("main.game_config")
local ID_YANDEX_METRICA = ____game_config.ID_YANDEX_METRICA
local ____System = require("modules.System")
local register_system = ____System.register_system
local ____Log = require("modules.Log")
local register_log = ____Log.register_log
local ____EventBus = require("modules.EventBus")
local register_event_bus = ____EventBus.register_event_bus
local ____Storage = require("modules.Storage")
local register_storage = ____Storage.register_storage
local ____GameStorage = require("modules.GameStorage")
local register_game_storage = ____GameStorage.register_game_storage
local ____Sound = require("modules.Sound")
local register_sound = ____Sound.register_sound
local ____Lang = require("modules.Lang")
local register_lang = ____Lang.register_lang
local ____Scene = require("modules.Scene")
local register_scene = ____Scene.register_scene
local ____Ads = require("modules.Ads")
local register_ads = ____Ads.register_ads
local ____Rate = require("modules.Rate")
local register_rate = ____Rate.register_rate
local ____Metrica = require("modules.Metrica")
local register_metrica = ____Metrica.register_metrica
local ____Camera = require("modules.Camera")
local register_camera = ____Camera.register_camera
local ____HtmlBridge = require("modules.HtmlBridge")
local register_html_bridge = ____HtmlBridge.register_html_bridge
function ManagerModule()
    local register_modules, check_ready, send_raw, send, _is_ready, MANAGER_ID
    function register_modules(callback_ready)
        register_metrica()
        register_sound()
        register_lang()
        register_scene()
        register_camera()
        register_ads()
        register_rate()
        Metrica.init(ID_YANDEX_METRICA)
        check_ready(callback_ready)
    end
    function check_ready(callback_ready)
        local id_timer
        id_timer = timer.delay(
            0.1,
            true,
            function()
                if Ads.is_ready() then
                    timer.cancel(id_timer)
                    _is_ready = true
                    log("All Managers ready ver: " .. sys.get_config("project.version"))
                    send("MANAGER_READY")
                    if callback_ready then
                        callback_ready()
                    end
                end
            end
        )
    end
    function send_raw(message_id, message_data, receiver)
        if receiver == nil then
            receiver = MANAGER_ID
        end
        msg.post(receiver, message_id, message_data)
    end
    function send(message_id, message_data, receiver)
        if receiver == nil then
            receiver = MANAGER_ID
        end
        send_raw(message_id, message_data, receiver)
    end
    _is_ready = false
    MANAGER_ID = "main:/manager"
    local UI_ID = "/ui#game"
    local LOGIC_ID = "/game_logic#game"
    local VIEW_ID = "/game_view#view"
    local function init(callback_ready, use_custom_storage_key)
        if use_custom_storage_key == nil then
            use_custom_storage_key = false
        end
        math.randomseed(socket.gettime())
        register_system()
        register_log()
        register_storage(use_custom_storage_key)
        register_game_storage()
        register_event_bus()
        if System.platform == "HTML5" then
            register_html_bridge()
            HtmlBridge.init(function() return register_modules(callback_ready) end)
        else
            register_modules(callback_ready)
        end
    end
    local function is_ready()
        return _is_ready
    end
    local function send_raw_ui(message_id, message_data, receiver)
        if receiver == nil then
            receiver = UI_ID
        end
        send_raw(message_id, message_data, receiver)
    end
    local function send_raw_view(message_id, message_data, receiver)
        if receiver == nil then
            receiver = VIEW_ID
        end
        send_raw(message_id, message_data, receiver)
    end
    local function send_view(message_id, message_data, receiver)
        if receiver == nil then
            receiver = VIEW_ID
        end
        send_raw(message_id, message_data, receiver)
    end
    local function send_raw_game(message_id, message_data, receiver)
        if receiver == nil then
            receiver = LOGIC_ID
        end
        send_raw(message_id, message_data, receiver)
    end
    local function send_game(message_id, message_data, receiver)
        if receiver == nil then
            receiver = LOGIC_ID
        end
        send_raw(message_id, message_data, receiver)
    end
    local function on_message(_this, message_id, message, sender)
        EventBus.on_message(_this, message_id, message, sender)
    end
    local function on_message_main(_this, message_id, message, sender)
        if _is_ready then
            Scene._on_message(_this, message_id, message, sender)
            Sound._on_message(_this, message_id, message, sender)
            Rate._on_message(_this, message_id, message, sender)
        end
    end
    local function init_script()
        Lang.apply()
        EventBus.on(
            "ON_APPLY_CUSTOM_LANG",
            function() return Lang.apply() end
        )
    end
    local function final_script()
        EventBus.off_all_current_script()
    end
    return {
        init = init,
        on_message_main = on_message_main,
        on_message = on_message,
        is_ready = is_ready,
        init_script = init_script,
        final_script = final_script,
        send = send,
        send_raw = send_raw,
        send_game = send_game,
        send_raw_game = send_raw_game,
        send_view = send_view,
        send_raw_view = send_raw_view,
        send_raw_ui = send_raw_ui,
        MANAGER_ID = MANAGER_ID,
        UI_ID = UI_ID,
        LOGIC_ID = LOGIC_ID,
        VIEW_ID = VIEW_ID
    }
end
local hashed_keys = {}
local function _to_hash(key)
    if hashed_keys[key] ~= nil then
        return hashed_keys[key]
    end
    hashed_keys[key] = hash(key)
    return hashed_keys[key]
end
function ____exports.register_manager()
    _G.Manager = ManagerModule()
    _G.to_hash = _to_hash
    _G.ID_MESSAGES = _ID_MESSAGES
end
return ____exports
